uh50_api/uh50.py

```diff
@@ -2,70 +2,300 @@
+    'flowrate_max_previous_year_m3ph': r'6.33\*01\((.*?)\*m3ph\)',
+    'flow_temperature_max_c': r'9.4\((.*?)\*C',
+    'return_temperature_max_c': r'9.4\(.*?\*C&(.*?)\*C',
+    'flow_temperature_max_previous_year_c': r'9.4\*01\((.*?)\*C',
+    'return_temperature_max_previous_year_c': r'9.4\*01\(.*?\*C&(.*?)\*C',
+    'operating_hours': r'6.31\((.*?)\*h\)',
+    'fault_hours': r'6.32\((.*?)\*h\)',
+    'fault_hours_previous_year': r'6.32\*01\((.*?)\*h\)',
+    'yearly_set_day': r'6.36\((.*?)\)',
+    'monthly_set_day': r'6.36\*02\((.*?)\)',
+    'meter_date_time': r'9.36\((.*?)\)',
+    'measuring_range_m3ph': r'9.24\((.*?)\*m3ph\)',
+    'settings_and_firmware': r'9.1\((.*?)\)',
+    'flow_hours': r'9.31\((.*?)\*h\)',
+}
 
 def find_ports():
     "Returns the available ports"
     return serial.tools.list_ports.comports()
 
-def _search_data(data):
-    match = re.search( r'6.8\((.*)\*GJ\)6.26\((.*)\*m3\)9.21\(66153690\)', str(data), re.M|re.I)
-    if match: 
-        return match.group(1),match.group(2)
-
-    raise Exception("GJ and m3 values not found")
-
-def _connect_serial(port) -> Serial:
-    return serial.Serial(port,
-                        baudrate=300,
-                        bytesize=serial.SEVENBITS,
-                        parity=serial.PARITY_EVEN,
-                        stopbits=serial.STOPBITS_TWO,
-                        timeout=1,
-                        xonxoff=0,
-                        rtscts=0
-                        )
-
-def _wake_up(conn) -> str:
-    # Waking up should be done at 300 baud
-    # Sending /?!
-    conn.write(b"\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x2F\x3F\x21\x0D\x0A")
-    ir_command='/?!\x0D\x0A'
-    conn.write(ir_command.encode('utf-8'))
-    conn.flush()   
-    return conn.readline().decode('utf-8')[1:9]   # Read at 300 baud, this gives us the typenr
-
-def _get_data(conn):
-    #TODO make this asyc
-    conn.baudrate=2400 # Now switch to 2400 BAUD. This could be different for other models. Let me know if you experience problems.
-    return conn.readline().decode('utf-8') # Reading just one line, because that's where we know the data is.
+class UH50:
+    """Class for reading the UH50 on the specified port"""
+
+    def __init__(self, port: str) -> None:
+        self.port = port
+
+    def update(self) -> None:
+        "Reads the UH50 on the specified port, searching for info on GJ and m3"
+        with self._connect_serial() as conn:
+            self.model = self._wake_up(conn)
+
+            # checking if we can read the model (eg. 'LUGCUH50')
+            if not(self.model):
+                raise Exception('No model could be read')
+            
+            self._full_response = self._get_data(conn)
+            self._gj, self._m3 = self._search_data(self._full_response)
+
+    def update_dummy(self) -> None:
```

Comment:
Kan me niet voorstellen dat deze nog moet blijven bestaan. Hij wordt wel getest, maar wat test je daar eigenlijk ?

Reply:
Ik snap wat je bedoelt. Dit heeft inderdaad een testfunctie, maar ook voor de gebruiker van de package. Het aanroepen van de heat meter kost batterij, dus er moet manier zijn om de integratie te maken en debuggen zonder de batterij telkens aan  te spreken (of uberhaupt zonder seriele connectie). Wat is hier een beter alternatief voor? Een parameter meegeven aan de update?

Reply:
De random hierin is overigens zodat het ook wat doet, visueel. Maar misschien is dat niet zo belangrijk voor de integratie. Ik zat ook te denken om de dummy_reponse uit te lezen en terug te geven, maar dan lopen testcase productie wel vreemd door elkaar. Wat is wijsheid ...

Reply:
Als je het parsen van die response hebt losgetrokken in een eigen class, dan kan je zo'n class aanmaken met een response die niet uit de dure UH50 call komt. Dan kan je daarmee verder testen. Is dat wellicht een oplossing voor het porbleem dat je beschrijft?

Alles wat een gebruiker van deze code zou moeten willen testen is toch juist de connectie met zijn/haar heatmeter? 

Het zou kunnen dat ik de usecase nog niet helemaal begrijp hoor. Kunnen we wellicht even over bellen.

Reply:
Deze package is pas de helft. De integratie in HA is waar het om gaat. Daar zit ook nog de nodige code. Ik wil tijdens die integratie de resultaten in HA kunnen zien, zonder echte UH50 calls.

Reply:
Ok, die snap ik. Je hebt hier last van dat er teveel dingen tegelijk gebeuren. Het lezen van de response en het parsen zit hier door elkaar. Het leven wordt beter als je die twee dingen uit elkaar trekt. Een class die een response uit de uh50 haalt, een class die een response kan parsen en een class die beiden aan elkaar verbindt. Dan kan je voor testdoeleinden een file based variant maken die de response niet uit de uh50, maar uit een file op disk leest. Voor testdoeleinden kan je dan de uh50 implementatie verwisselen met die file based implementatie